### ISSUE TITLE
Remove Reader interface

### ISSUE THREAD

**Issue by bruth** [chop-dbhi/origins#140]
`Reader` is modeled after `io.Reader` which fills a passed buffer of facts and returns the number of facts read or an error. After implementing and using the various APIs, the `Reader` interface does not appear to be used frequently if at all. The dominant interface being used is `Iterator`, which aligns well with the log-based nature of facts.

A quick grep of the codebase shows the `Reader` interface is only used in three places:

- `origins.CSVReader`
- `origins.Buffer`
- `testutil.FactGen`

@l-k- Thougts on this? This will enable us to clean up some of the ambiguous private auxillary types that are in use like `logView` and `csvReader`, etc.

**Comment by l-k-** [comment]
That's what I thought as well. Should I give it a shot?  I am curious what you meant in your original [comment](https://github.com/chop-dbhi/origins/pull/136#discussion_r32388722) when you said "A simpler solution may be to expose those types." But I'll ask you tomorrow, got sidetracked with the benchmarking issue today.

**Comment by bruth** [comment]
I should have a PR soon for this one. It was straightforward and prevents having to support `Reader` future changes.

**Comment by l-k-** [comment]
sounds good

**Comment by bruth** [comment]
I did some more research for the idiomatic *iterator* interface. I am not really satisfied with the current `Iterator`'s interface. Having to break out of a `for` loop and do error checking each iteration, etc. feels cumbersome.

I was pointed to [`bufio.Scanner`](http://golang.org/pkg/bufio/#Scanner) as a possible interface to model after. The role of a `Scanner` is to lazily *scan* tokens from an `io.Reader` (commonly a file). I think it is wise to break free from the `Reader` interface and define one that better matches the primary interface in Origins which is essentially a log of facts. We currently have `view.Log` (which I will be opening an issue for later), but it would be more valuable to define a `Log` interface similar to `Scanner`. My proposed interface is:

```go
type Log interface {
	Scan() bool
	Next() *Fact
	Err() error
}
```

The usage of it would be:

```go
// Scan *the log* until no more facts are available or an error occurred.
for l.Scan() {
    f := l.Next()
    // Do something with the fact
}

// Check if an error occurred while scanning.
if l.Err() != nil {
    // Do something with the error.
}
```

The `Scan()` method will return false if there are no more facts or an error occurs which is a more straightforward interface, basically "scan/read until you can't".

Two things I am not settled on:

- `Log` for the interface name
- `Scan` for the name of the method that toggles the iteration

**Comment by bruth** [comment]
> Two things I am not settled on:

Thinking about this more, the interface should be called the more idiomatic `Scanner`, not `Log`. I am going to put together a PR with these changes for review.

**Comment by bruth** [comment]
After talking to @l-k-, the decided interface is:

```go
type Iterator interface {
    Next() *Fact
    Err() error
}
```

with a usage of:

```go
// Next advances the pointer and returns the next fact in the stream
// if available. If an error occurs, nil will also be returned. `Err` can be
// called to get the error if one occurred.
for f := it.Next(); f != nil {
    // Do something with the fact...
}

if err := it.Err(); err != nil {
    // Handle error
}
```

One other note while implementing the `Iterator` interface is that `io.EOF` should no longer be used to denote the *end of a stream* and should not propagate up as a real error. If it is an unexpected `EOF` then the implementation should handle this and return a more informative error.

**Comment by bruth** [comment]
Fixed in #141
